### ISSUE TITLE
Babel-SAPBERT bug in how the count parameter affects the search results

### ISSUE THREAD

**Issue by gaurav** [RENCI-NER/nemo-serve#22]
If you search for:

```
{
  "text": "asthma",
  "model_name": "sapbert",
  "count": 100,
  "args": {"bl_type": "Disease"}
}
```

we find the MONDO asthma at a distance of 0.11 -- great! But, if we remove the Biolink type:

```
{
  "text": "asthma",
  "model_name": "sapbert",
  "count": 100
}
```

our top results are large distance, unrelated chemicals:

```
[
  {
    "name": "2-(2-methylphenoxy)-N-{2-[(2-morpholin-4-yl-2-oxoethyl)thio]-1,3-benzothiazol-6-yl}acetamide",
    "curie": "PUBCHEM.COMPOUND:16015494",
    "category": "biolink:SmallMolecule",
    "score": "0.556742727757"
  },
  {
    "name": "2-(2-chlorophenoxy)-N-{2-[(2-morpholin-4-yl-2-oxoethyl)thio]-1,3-benzothiazol-6-yl}acetamide",
    "curie": "PUBCHEM.COMPOUND:16015518",
    "category": "biolink:SmallMolecule",
    "score": "0.558322906494"
  },
  {
    "name": "2-(2,5-dimethylphenoxy)-N-{2-[(4-methyl-1-piperazinyl)carbonyl]phenyl}acetamide",
    "curie": "PUBCHEM.COMPOUND:4943673",
    "category": "biolink:SmallMolecule",
    "score": "0.558954298496"
  },
```

Probably a bug in the HNSW settings described at https://redis.io/docs/stack/search/reference/vectors/?

**Comment by gaurav** [comment]
This bug can be fixed by increasing the count, i.e. it occurs when count=100 but not when count=1000. No, really.

The same applies to other queries as well, e.g. `diabetes` works correctly when count=100, but you end up with a bunch of random chemicals when count=10. Increasing count doesn't actually increase the number of results (which max out at 10) or the time taken for a response.

**Comment by gaurav** [comment]
@YaphetKG Do you have any ideas what might be causing this?
@mbacon-renci Any chance you could take a look at this when you have time next?

**Comment by YaphetKG** [comment]
Yeah i think according to the docs above , EPSILON and EP_RUNTIME, parameters affect the accuracy by increasing the search space where HSNW is performed. 
But the BL filter does decrease the search space hence returning the more accurate results. 

The caveat with adjusting EPSILON and EP_RUNTIME, is that performance may be affected but, we can test out different parameters to see what works best. If we want accurate results with out specifying BL types

**Comment by mbacon-renci** [comment]
@gaurav Does Yaphet's answer cover this, or do I need to do more research? Are there instances where we need better results without bl_type or can we just say, "results will suck without bl_type?"

**Comment by gaurav** [comment]
@mbacon-renci Actually, we see this bug even with bl_type. For example, the following query doesn't return any results on https://babel-sapbert.apps.renci.org/docs#/default/annotate_text_annotate__post:

```json
{
  "text": "hypertension",
  "model_name": "sapbert",
  "count": 10,
  "args": {"bl_type": "PhenotypicFeature"}
}
```

But the following does find HP:0032263 ("Increased blood pressure") as well we a bunch of UMLS identifiers:

```json
{
  "text": "hypertension",
  "model_name": "sapbert",
  "count": 1000,
  "args": {"bl_type": "PhenotypicFeature"}
}
```

**Comment by gaurav** [comment]
Count might be mislabeled here: it's talking about wide your search is. Lets you control how quickly you want your search to return.

**Comment by bpow** [comment]
Apropos discussion on the call today, If I'm looking at the right branch, `num_candidates` is limiting the max meaningful `count` parameter.

```
            "knn": {
                "field": "embedding",
                "query_vector": query_vector,
                "k": top_n,
                # defaulting to consider 10000 docs as candidates,
                "num_candidates": 10000
            },
```

[elastic API docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/knn-search-api.html) say that 10,000 is as high as we can go with `num_candidates`, anyway.

So this doesn't answer the main questions here, but says why we can't set count to a million and get any results back.
